### ISSUE TITLE
Behavior of Create[Bit, Byte, Word, DWord, QWord]Field() differs from Windows

### ISSUE THREAD

**Issue by Wer-Wolf** [acpica/acpica#953]
After encountering ACPI errors while developing a WMI driver for linux, i had to find out that the behavior of:

- `CreateField()`
- `CreateBitField()`
- `CreateByteField()`
- `CreateWordField()`
- `CreateDWordField()`
- `CreateQWordField()`

differs from Windows in major ways:

1. Using a undersized buffer for creating buffer fields does not immediately throw an error. The error only happens when the buffer field is ultimately accessed later. ACPICA however throws an error immediately. 
2. When assigning a new buffer to the variable used to create a buffer field, previously created buffer fields are updated to point to the new buffer. ACPICA however does not update previously created buffer fields.

I tested this using two SSDT tables, `eval.asl` and `change.asl`.

`eval.asl` throws errors when being executed using `acpiexec`, but returns "TEST" when executed by `Acpi.sys`.
`change.asl` returns the content of the first buffer when being executed using `acpiexec`, but returns the content of the second buffer when executed by `Acpi.sys`.

I suspect that Windows creates a mapping to the underlying variable containing the buffer and not the buffer itself, and this mapping is only validated upon access.

This difference in behavior causes various ACPI error under linux since some vendors depend on the Windows behavior by creating various buffer fields first and only after that check if they can be accessed.

**Comment by d-tatianin** [comment]
This is very interesting. So what happens if `Local0` is changed to an object that's not suitable for a buffer field? E.g. `Local0 = Package { 1 }`?

**Comment by Wer-Wolf** [comment]
I will check, give me a minute

**Comment by d-tatianin** [comment]
Looks like uACPI is also not following Acpi.sys correctly in this case (but rather ACPICA), but luckily should be a trivial fix for both.
```
[uACPI][ERROR] Invalid buffer field: bits [255..256], buffer size is 4 bytes
[uACPI][ERROR] aborting method invocation due to previous error: out of bounds AML index
[uACPI][ERROR]     #0 in \_SB_.TF1_._HID()
```

**Comment by Wer-Wolf** [comment]
Executing this code still return `0x0` under Windows.

    Local0 = Buffer (4) { 0xad, 0xad, 0xad, 0xad}
    CreateBitField(Local0, 0x0, BIT)
    Local0 = Package { 1 }
    Return (ToHexString(BIT))

**Comment by d-tatianin** [comment]
> Executing this code still return `0x0` under Windows.
> 
> ```
> Local0 = Buffer (4) { 0xad, 0xad, 0xad, 0xad}
> CreateBitField(Local0, 0x0, BIT)
> Local0 = Package { 1 }
> Return (ToHexString(BIT))
> ```

I wonder where it's getting that value from exactly...?

**Comment by Wer-Wolf** [comment]
I will check if `0x0` is some sort of fallback value.

**Comment by d-tatianin** [comment]
Could be that it just blindly dereferences whatever object is still pointed to by Local0 and treats it as a buffer

**Comment by Wer-Wolf** [comment]
[invalid.asl.txt](https://github.com/acpica/acpica/files/15287485/invalid.asl.txt)

The results are:

- TES1: 0x0
- TES2: 0x54
- TES3: 0x18
- TES4: 0x18

I suspect that `Acpi.sys` does not check whether the object is a buffer and just blindly treats it as such.

**Comment by d-tatianin** [comment]
Looks like it should be trivial to crash it this way. Probably not the behavior you would want to be compatible with 😃

**Comment by Wer-Wolf** [comment]
Yes, if we adopt this behavior from windows, then we should enhance it with some error checking.

**Comment by Wer-Wolf** [comment]
Wait, let me try something.

**Comment by d-tatianin** [comment]
I suspect the same thing would happen for
```python
Name (X, Buffer { 0xFF })
CreateByteField(X, 0x0, BYTE)
CopyObject(Buffer { 0x00 }, X)
```

**Comment by d-tatianin** [comment]
@Wer-Wolf Does this mean that the initial type of the object being used as the field is also not checked, or is checked once at creation time but then ignored afterwards?

**Comment by Wer-Wolf** [comment]
I just checked, and it seems that the initial type of the object being used as the field is checked once at creation time (using a Package instead of a Buffer for the initial object type leads to a BSOD).
